torch/functional.py

```diff
@@ -1069,14 +1077,32 @@ def norm(input, p="fro", dim=None, keepdim=False, out=None, dtype=None):  # noqa
 
     ndim = input.dim()
 
-
     # catch default case
-    if dim is None and out is None and dtype is None and p is not None:
-        if isinstance(p, str):
-            if p == "fro":
-                return _VF.frobenius_norm(input)
-        if not isinstance(p, str):
-            return _VF.norm(input, p)
+    if dim is None and out is None and dtype is None:
+        if p is None:
+            return _VF.frobenius_norm(input, dim=(), keepdim=keepdim)
+
+        elif isinstance(p, str):
+            if p == 'fro':
+                if ndim != 2:
+                    raise RuntimeError('p="%s" can only be set if len(dim) == 2' % (p))
+                return _VF.frobenius_norm(input, dim=(), keepdim=keepdim)
+            elif p == 'nuc':
+                if ndim != 2:
+                    raise RuntimeError('p="%s" can only be set if len(dim) == 2' % (p))
+                return _VF.nuclear_norm(input, keepdim=keepdim)
+            else:
+                raise RuntimeError('Invalid norm order for vectors: %s' % p)
+
+        else:
+            if ndim == 1:
+                return _VF.norm(input, p, dim=(), keepdim=keepdim)
```

Comment:
I'm having a new issue with JIT after I fixed support for `keepdim=True`:

```
$ python test/test_jit.py TestJitGeneratedAutogradCPU.test_norm_neg_2_cpu
...
File "/opt/conda/lib/python3.6/site-packages/torch/functional.py", line 1099, in norm
    return _VF.norm(input, p, dim=(), keepdim=keepdim)
RuntimeError: Tracing a list of arbitrary type is currently not supported!
```

Originally, this line was `return _VF.norm(input, p)`, but when I added the keepdim argument to this call of `_VF.norm()`, I also had to add the `dim` argument since that argument is required for all available overloads of `_VF.norm()` that include the `keepdim` arg. But I needed `dim` to be the default argument of an empty tuple/list, for correct behavior. I'm seeing the same issue with `_VF._norm_matrix()` as well

Since JIT cannot trace an empty list, I think my only option is to modify the schemas to support calling with the arguments `(input, p, keepdim=keepdim)`. That is, unless there's something I can use in place of `dim=()` to tell JIT that this should be an empty list of ints, rather than just an arbitrary type.

Reply:
Oh woops, I just realized the solution is very easy. I just have to call with `dim=0` in the vector norm case, and `dim=(0,1)` in the matrix norm case.